Binaries/DafnyRuntime.cs

```diff
@@ -1005,6 +1000,79 @@ public class Sequence<T>
       return Subsequence((long)lo, (long)hi);
     }
   }
+  internal class ArraySequence<T> : Sequence<T> {
+    private readonly T[] elmts;
+
+    internal ArraySequence(T[] ee) {
+      elmts = ee;
+    }
+    public override T[] Elements {
+      get {
+        return elmts;
+      }
+    }
+    public override long LongCount {
+      get {
+        return elmts.LongLength;
+      }
+    }
+  }
+  internal class ConcatSequence<T> : Sequence<T> {
+    private Sequence<T> left, right;
+    private T[] elmts = null;
+    private readonly long count;
+
+    internal ConcatSequence(Sequence<T> left, Sequence<T> right) {
+      this.left = left;
+      this.right = right;
+      this.count = left.LongCount + right.LongCount;
+    }
+
+    public override T[] Elements {
+      get {
+        if (elmts == null) {
+          elmts = ComputeElements();
+          // We don't need the original sequences anymore; let them be
+          // garbage-collected
+          left = null;
+          right = null;
+        }
+        return elmts;
+      }
+    }
+
+    public override long LongCount {
+      get {
+        return count;
+      }
+    }
+
+    private T[] ComputeElements() {
+      // Traverse the tree formed by all descendants which are ConcatSequences
+
+      var list = new List<T>();
```

Comment:
Since the final length is known, this `List<T>` can be replaced by a `T[]`. This saves the list from having to resize (which would also be prevented by giving an initial capacity to the `List<T>` constructor), and it saves the final copy made by `ToArray` on line 1073. The cost is low: replace the line-1063 `AddRange` with copying that keeps track of indices.

Reply:
Good idea, done.